File: src/features/organizations/data/models/organization_model.ts
/**
 * Negocio Flex - Organization Data Model (Fase 4)
 */

import {
  OrganizationEntity,
  BusinessType,
  OrganizationStatus,
  OrganizationRole,
  OrgBrandingEntity,
  OrgModulesConfigEntity,
  OrganizationSettingsEntity,
} from '../../domain/entities/organization_entity';

export class OrganizationModel implements OrganizationEntity {
  constructor(
    public readonly id: string,
    public readonly name: string,
    public readonly slug: string,
    public readonly businessType: BusinessType,
    public readonly status: OrganizationStatus,
    public readonly createdBy: string,
    public readonly createdAt: string,
    public readonly updatedAt: string,
    public readonly currency: string,
    public readonly branding: OrgBrandingEntity,
    public readonly modules: OrgModulesConfigEntity,
    public readonly description?: string,
    public readonly phone?: string,
    public readonly email?: string,
    public readonly address?: string,
    public readonly settings?: OrganizationSettingsEntity,
    public readonly currentUserRole?: OrganizationRole,
    public readonly memberCount?: number,
  ) {}

  static fromJson(json: any): OrganizationModel {
    const rawBranding = json.branding || {};
    const rawModules = json.modules || {};
    const rawSettings = Array.isArray(json.organization_settings)
      ? json.organization_settings[0] || {}
      : json.organization_settings || json.settings || {};

    const activeModules = rawSettings.active_modules || rawModules || {};

    return new OrganizationModel(
      json.id,
      json.name || 'Mi Negocio',
      json.slug || (json.name ? json.name.toLowerCase().replace(/[^a-z0-9]+/g, '-') : 'mi-negocio'),
      (json.business_type || json.businessType || 'restaurant') as BusinessType,
      (json.status || (json.is_active === false ? 'inactive' : 'active')) as OrganizationStatus,
      json.created_by || json.createdBy || json.owner_id || json.ownerId || '',
      json.created_at || json.createdAt || new Date().toISOString(),
      json.updated_at || json.updatedAt || json.created_at || new Date().toISOString(),
      json.currency || rawSettings.currency || 'S/',
      {
        primaryColor: rawSettings.primary_color || rawBranding.primaryColor || rawBranding.primary_color || '#4F46E5',
        secondaryColor: rawSettings.secondary_color || rawBranding.secondaryColor || rawBranding.secondary_color || '#064E3B',
        logoUrl: rawSettings.logo_url || rawBranding.logoUrl || rawBranding.logo_url || 'https://images.unsplash.com/photo-1578985545062-69928b1d9587?w=150',
        bannerUrl: rawSettings.cover_url || rawBranding.bannerUrl || rawBranding.banner_url,
        fontName: rawBranding.fontName || 'Inter',
        slogan: rawSettings.slogan || rawBranding.slogan || 'Calidad y servicio garantizado',
      },
      {
        enableProducts: activeModules.products ?? rawModules.enableProducts ?? rawModules.enable_products ?? true,
        enableServices: activeModules.services ?? rawModules.enableServices ?? rawModules.enable_services ?? false,
        enableAppointments: activeModules.appointments ?? rawModules.enableAppointments ?? rawModules.enable_appointments ?? false,
        enableInventory: activeModules.inventory ?? rawModules.enableInventory ?? rawModules.enable_inventory ?? true,
        enableOrders: activeModules.orders ?? rawModules.enableOrders ?? rawModules.enable_orders ?? true,
        enableWhatsappCheckout: activeModules.whatsapp ?? rawModules.enableWhatsappCheckout ?? rawModules.enable_whatsapp_checkout ?? true,
        enableStaffManagement: activeModules.staff ?? rawModules.enableStaffManagement ?? rawModules.enable_staff_management ?? false,
        enableReviews: activeModules.reviews ?? rawModules.enableReviews ?? rawModules.enable_reviews ?? true,
      },
      json.description || '',
      rawSettings.phone || json.phone,
      rawSettings.email || json.email,
      rawSettings.address || json.address,
      rawSettings.id
        ? {
            id: rawSettings.id,
            organizationId: rawSettings.organization_id || json.id,
            logoUrl: rawSettings.logo_url,
            coverUrl: rawSettings.cover_url,
            primaryColor: rawSettings.primary_color || '#4F46E5',
            secondaryColor: rawSettings.secondary_color || '#064E3B',
            accentColor: rawSettings.accent_color || '#F59E0B',
            textColor: rawSettings.text_color || '#111827',
            address: rawSettings.address,
            phone: rawSettings.phone,
            whatsappNumber: rawSettings.whatsapp_number,
            whatsappMessage: rawSettings.whatsapp_message,
            email: rawSettings.email,
            instagramUrl: rawSettings.instagram_url,
            facebookUrl: rawSettings.facebook_url,
            tiktokUrl: rawSettings.tiktok_url,
            youtubeUrl: rawSettings.youtube_url,
            websiteUrl: rawSettings.website_url,
            currency: rawSettings.currency || 'S/',
            slogan: rawSettings.slogan,
            activeModules: typeof activeModules === 'object' ? activeModules : undefined,
            language: rawSettings.language || 'es',
            timezone: rawSettings.timezone || 'America/Lima',
            createdAt: rawSettings.created_at || new Date().toISOString(),
            updatedAt: rawSettings.updated_at || new Date().toISOString(),
          }
        : undefined,
      (json.currentUserRole || json.role || 'owner') as OrganizationRole,
      json.memberCount ?? json.member_count ?? 1,
    );
  }

  toJson(): Record<string, any> {
    return {
      id: this.id,
      name: this.name,
      slug: this.slug,
      business_type: this.businessType,
      status: this.status,
      created_by: this.createdBy,
      created_at: this.createdAt,
      updated_at: this.updatedAt,
      description: this.description,
      phone: this.phone,
      email: this.email,
      address: this.address,
      currency: this.currency,
      branding: {
        primary_color: this.branding.primaryColor,
        secondary_color: this.branding.secondaryColor,
        logo_url: this.branding.logoUrl,
        banner_url: this.branding.bannerUrl,
        font_name: this.branding.fontName,
        slogan: this.branding.slogan,
      },
      modules: {
        enable_products: this.modules.enableProducts,
        enable_services: this.modules.enableServices,
        enable_appointments: this.modules.enableAppointments,
        enable_inventory: this.modules.enableInventory,
        enable_orders: this.modules.enableOrders,
        enable_whatsapp_checkout: this.modules.enableWhatsappCheckout,
        enable_staff_management: this.modules.enableStaffManagement,
        enable_reviews: this.modules.enableReviews,
      },
      currentUserRole: this.currentUserRole,
      member_count: this.memberCount,
    };
  }
}

File: flutter_project/lib/features/organizations/data/models/organization_model.dart
/// Negocio Flex - Organization Data Model (Flutter / Dart)

import '../../domain/entities/organization_entity.dart';

class OrganizationModel extends OrganizationEntity {
  const OrganizationModel({
    required super.id,
    required super.name,
    required super.slug,
    required super.businessType,
    super.status,
    required super.createdBy,
    required super.createdAt,
    required super.updatedAt,
    super.description,
    super.phone,
    super.email,
    super.address,
    super.currency,
    super.branding,
    super.modules,
    super.currentUserRole,
    super.memberCount,
  });

  factory OrganizationModel.fromJson(Map<String, dynamic> json) {
    final rawBranding = json['branding'] as Map<String, dynamic>? ?? {};
    final rawModules = json['modules'] as Map<String, dynamic>? ?? {};

    BusinessType parseBusinessType(String? type) {
      switch (type) {
        case 'restaurant': return BusinessType.restaurant;
        case 'salon': return BusinessType.salon;
        case 'gym': return BusinessType.gym;
        case 'store': return BusinessType.store;
        case 'professional': return BusinessType.professional;
        default: return BusinessType.other;
      }
    }

    OrganizationRole? parseRole(String? role) {
      switch (role) {
        case 'owner': return OrganizationRole.owner;
        case 'admin': return OrganizationRole.admin;
        case 'staff': return OrganizationRole.staff;
        default: return null;
      }
    }

    return OrganizationModel(
      id: json['id'] ?? '',
      name: json['name'] ?? 'Mi Negocio',
      slug: json['slug'] ?? 'mi-negocio',
      businessType: parseBusinessType(json['business_type'] ?? json['businessType']),
      status: json['status'] == 'inactive' ? OrganizationStatus.inactive : OrganizationStatus.active,
      createdBy: json['created_by'] ?? json['createdBy'] ?? '',
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : DateTime.now(),
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : DateTime.now(),
      description: json['description'],
      phone: json['phone'],
      email: json['email'],
      address: json['address'],
      currency: json['currency'] ?? 'S/',
      branding: OrgBrandingEntity(
        primaryColor: rawBranding['primary_color'] ?? rawBranding['primaryColor'] ?? '#4F46E5',
        secondaryColor: rawBranding['secondary_color'] ?? rawBranding['secondaryColor'] ?? '#0D9488',
        logoUrl: rawBranding['logo_url'] ?? rawBranding['logoUrl'],
        bannerUrl: rawBranding['banner_url'] ?? rawBranding['bannerUrl'],
        slogan: rawBranding['slogan'],
      ),
      modules: OrgModulesConfigEntity(
        enableProducts: rawModules['enable_products'] ?? rawModules['enableProducts'] ?? true,
        enableServices: rawModules['enable_services'] ?? rawModules['enableServices'] ?? true,
        enableAppointments: rawModules['enable_appointments'] ?? rawModules['enableAppointments'] ?? false,
        enableInventory: rawModules['enable_inventory'] ?? rawModules['enableInventory'] ?? true,
        enableOrders: rawModules['enable_orders'] ?? rawModules['enableOrders'] ?? true,
        enableWhatsappCheckout: rawModules['enable_whatsapp_checkout'] ?? rawModules['enableWhatsappCheckout'] ?? true,
        enableStaffManagement: rawModules['enable_staff_management'] ?? rawModules['enableStaffManagement'] ?? true,
        enableReviews: rawModules['enable_reviews'] ?? rawModules['enableReviews'] ?? true,
      ),
      currentUserRole: parseRole(json['currentUserRole'] ?? json['role']),
      memberCount: json['member_count'] ?? json['memberCount'] ?? 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'slug': slug,
      'business_type': businessType.name,
      'status': status.name,
      'created_by': createdBy,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'description': description,
      'phone': phone,
      'email': email,
      'address': address,
      'currency': currency,
      'member_count': memberCount,
    };
  }
}
